lib/Maskings/Maskings.cpp

```diff
@@ -0,0 +1,309 @@
+      return VPackValue(VPackValueType::Null);
+    }
+  } else {
+    if (data.isBool()) {
+      return func->mask(data.getBool());
+    } else if (data.isString()) {
+      velocypack::ValueLength length;
+      char const* c = data.getString(length);
+      return func->mask(std::string(c, length), buffer);
+    } else if (data.isInteger()) {
+      return func->mask(data.getInt());
+    } else if (data.isDouble()) {
+      return func->mask(data.getDouble());
+    } else {
+      return VPackValue(VPackValueType::Null);
+    }
+  }
+
+  return VPackValue(xxxx);
+}
+
+void Maskings::addMaskedArray(Collection& collection, VPackBuilder& builder,
+                              std::vector<std::string>& path,
+                              VPackSlice const& data) {
+  for (auto const& entry : VPackArrayIterator(data)) {
+    if (entry.isObject()) {
+      VPackObjectBuilder ob(&builder);
+      addMaskedObject(collection, builder, path, entry);
+    } else if (entry.isArray()) {
+      VPackArrayBuilder ap(&builder);
+      addMaskedArray(collection, builder, path, entry);
+    } else {
+      std::string buffer;
+      builder.add(maskedItem(collection, path, buffer, entry));
+    }
+  }
+}
+
+void Maskings::addMaskedObject(Collection& collection, VPackBuilder& builder,
+                               std::vector<std::string>& path,
+                               VPackSlice const& data) {
+  for (auto const& entry : VPackObjectIterator(data, false)) {
+    std::string key = entry.key.copyString();
+    VPackSlice const& value = entry.value;
+
+    path.push_back(key);
+
+    if (value.isObject()) {
+      VPackObjectBuilder ob(&builder, key);
+      addMaskedObject(collection, builder, path, value);
+    } else if (value.isArray()) {
+      VPackArrayBuilder ap(&builder, key);
+      addMaskedArray(collection, builder, path, value);
+    } else {
+      std::string buffer;
+      builder.add(key, maskedItem(collection, path, buffer, value));
+    }
+
+    path.pop_back();
+  }
+}
+
+void Maskings::addMasked(Collection& collection, VPackBuilder& builder,
+                         VPackSlice const& data) {
+  if (!data.isObject()) {
+    return;
+  }
+
+  std::vector<std::string> path;
+  std::string dataStr("data");
+  VPackObjectBuilder ob(&builder, dataStr);
+
+  addMaskedObject(collection, builder, path, data);
+}
+
+void Maskings::addMasked(Collection& collection, basics::StringBuffer& data,
+                         VPackSlice const& slice) {
+  if (!slice.isObject()) {
+    return;
+  }
+
+  velocypack::StringRef dataStrRef("data");
+
+  VPackBuilder builder;
+
+  {
+    VPackObjectBuilder ob(&builder);
+
+    for (auto const& entry : VPackObjectIterator(slice, false)) {
+      velocypack::StringRef key = entry.key.stringRef();
+
+      if (key.equals(dataStrRef)) {
+        addMasked(collection, builder, entry.value);
+      } else {
+        builder.add(key, entry.value);
+      }
+    }
+  }
+
+  std::string masked = builder.toJson();
```

Comment:
we should not stringify the data here, but instead use a `VelocyPackDumper` (lib/Basics/VelocyPackDumper.h) instance to directly dump the JSON into the `StringBuffer` without any intermediate conversions.